Add global color tokens to color tokens tester page

lyzhan7 opened the pull request:
### Platforms Impacted
- [x] iOS
- [x] macOS
- [x] win32 (Office)
- [x] windows
- [x] android

### Description of changes

Commit 1: some updates to the existing alias color token tester logic
- Remove pickerContainer style - wasn’t being used
- Name/color/key was a little ambiguous to me- updated variable names (ex. colorName, colorValue)
- Moved logic to append the color hex value to the name into the ColorToken component
- Add warning if we try to show a color that doesn't have a definition

Commit 2/3: add global color tokens
- Generally copied how the alias color token swatch logic worked, main difference is that since global tokens don't change depending on the theme, I could move a lot of the logic into consts outside of the function component 
- If any of the logic is unclear let me know
- This section was added under a toggle button. Without it, the new section fails the iOS e2e tests - I think it's because there are over 500 global color tokens. When I tried to only render the first one, the test would pass.

Future work
- Will be changing the layout to be able to suit mobile a bit better (right now a lot of the long token names are getting truncated
- There are over 500 global shared colors so the page is pretty long now. Might consider adding collapsible sections or something so it's easier to find the right token, if anyone has any ideas let me know
- Known issue - switching between light/dark mode doesn't always update the alias color tokens. Will look into 

### Verification

iOS (*updated to include the new toggle button for global tokens)

https://user-images.githubusercontent.com/78454019/215227773-98769374-a9b9-4942-a204-b9106e6e5f38.mp4



macOS (*updated to include the new toggle button for global tokens)

https://user-images.githubusercontent.com/78454019/215227803-12a01691-a9b6-4839-bfb3-7c99c2cae193.mov



### Pull request checklist

This PR has considered (when applicable):
- [ ] Automated Tests
- [ ] Documentation and examples
- [ ] Keyboard Accessibility
- [ ] Voiceover
- [ ] Internationalization and Right-to-left Layouts

rurikoaraki commented on apps/fluent-tester/src/TestComponents/ColorTokens/ColorTokenTest.tsx:
```diff
@@ -29,52 +30,56 @@ const styles = StyleSheet.create({
         {...testProps(COLORTOKENS_TEST_COMPONENT)}
       />
-      <Text>{p.name}</Text>
+      <View>
```
nit: is there a particular reason for this View?

lyzhan7 replied:
Oh thanks for catching that, no I accidentally left that in when I was trying to figure out a way to get the text to wrap nicely on iOS, will remove

rurikoaraki commented on apps/fluent-tester/src/TestComponents/ColorTokens/ColorTokenTest.tsx:
```diff
@@ -83,15 +88,74 @@ const AliasTokensSwatchList: React.FunctionComponent = () => {
+  .reduce((accumulator, sharedColor) => accumulator.concat(sharedColor));
+
+const GlobalSharedColorTokensSwatchList: React.FunctionComponent = () => {
+  const renderSwatch = React.useCallback((item) => {
```
Seems like you could pull this out into a generic function, since the code is duplicated across several components and doesn't depend on anything but the input

lyzhan7 commented:
Not sure why the iOS/win32 e2e tests are failing the color tokens test (twice now). Looking into it

Saadnajmi commented on apps/E2E/src/ColorTokens/specs/ColorTokens.spec.ios.ts:
```diff
@@ -15,7 +15,7 @@ describe('Color Tokens Testing Initialization', function () {
     /* Click on component button to navigate to test page */
     await NavigateAppPage.clickAndGoToColorTokensPage();
-    await ColorTokenPageObject.waitForPageDisplayed(PAGE_TIMEOUT);
+    await ColorTokenPageObject.waitForPageDisplayed(PAGE_TIMEOUT * 2);
```
Should we modify the constant instead?

Saadnajmi commented on apps/fluent-tester/src/TestComponents/ColorTokens/ColorTokenTest.tsx:
```diff
@@ -29,52 +30,59 @@ const styles = StyleSheet.create({
-type ColorTokenProps = { color: ColorValue; name: string };
+type ColorTokenProps = { colorValue: ColorValue; colorName: string };
 const ColorToken: React.FunctionComponent<ColorTokenProps> = (p: ColorTokenProps) => {
+  if (p.colorValue === undefined) {
+    console.log('Color token ' + p.colorName + ' is undefined');
+  }
```
Should this be a yellow box warning?

lyzhan7 replied:
Which one is yellow box? I had this as console.warn at first, wanted to try this in case that fixed the e2e tests but since it didn't will probably go back to console.warn

Saadnajmi replied:
I think console.warn()

lyzhan7 replied:
Changed it back to console.warn

Saadnajmi commented on apps/fluent-tester/src/TestComponents/ColorTokens/ColorTokenTest.tsx:
```diff
@@ -83,15 +91,65 @@ const AliasTokensSwatchList: React.FunctionComponent = () => {
+const globalNeutralColorNamesSortedDarkToLight = Object.keys(globalTokens.color)
+  .filter((globalColorName) => globalColorName.includes('grey') || globalColorName === 'black' || globalColorName === 'white')
+  .sort((color1, color2) => {
+    const color1hex = globalTokens.color[color1];
```
Nit: colors aren't guaranteed to be a hex so maybe this should be rethought? Not a blocking comment.

lyzhan7 replied:
Oh I assumed they were. What else can the color value be?

lyzhan7 replied:
If this is still a concern feel free to reopen

Saadnajmi replied:
They can be a platformColor, or a Dynamic Color. Something that looks like:
`semantic: { 'windowBackgroundColor' }}` or `{light: 'hex', dark: 'hex'}

Saadnajmi replied:
Every platform can define their own types, so it's not worth assuming the structure.

samuelfreiberg commented on apps/fluent-tester/src/TestComponents/ColorTokens/ColorTokenTest.tsx:
```diff
@@ -100,7 +158,17 @@ const themeSections: TestSection[] = [
+  },
+  {
+    name: 'Global Color Tokens - Neutral Colors',
+    testID: COLORTOKEN_TESTPAGE,
```
These testID's are unnecessay. We only need one set to the first test section (in this case 'Alias Color Tokens', the two new ones don't need it

samuelfreiberg commented:
/azp run

lyzhan7 commented on yarn.lock:
```diff
@@ -1470,24 +1470,6 @@
   resolved "https://registry.yarnpkg.com/@fluentui-react-native/design-tokens-windows/-/design-tokens-windows-0.42.0.tgz#6b7ca87ed30a92634e72cbe8fc1280fd21db50b8"
   integrity sha512-hBPCikSDJ3Rgw+mAbTgnNB83dn3SeSKPo4XNli22tbV08fN8P5mEaEg8Uw0cbVMN3D138ykNeAaK/lZjO9KIqA==
 
-"@fluentui-react-native/framework@0.8.31":
```
will revert changes to this file

lyzhan7 commented:
| Commit | Passed ios e2e tests? |
| - | - |
| [Comment out global tokens tests to see if e2e tests pass](https://github.com/microsoft/fluentui-react-native/pull/2538/commits/b60c2fe0d0bd873eef0ca1fc4ead4ad49bf75e23) | passed |
| [Uncomment globalNeutralColors (global shared colors commented out)](https://github.com/microsoft/fluentui-react-native/pull/2538/commits/d6f32c9c7ebf29896c288d4dfa8d216b0ee2c0c8) | passed |
| [Uncomment global shared color tokens (comment out global neutral colors)](https://github.com/microsoft/fluentui-react-native/pull/2538/commits/865d3e913ac5940d66c34dea2b0ecf9e263f6f7d) | failed |
| [Uncomment everything, try global shared color test with just one color](https://github.com/microsoft/fluentui-react-native/pull/2538/commits/547f9e8ab5e5289422e6f4831ef08cc8c4b1200a) | failed |
| [Try commenting out inner view, logging globalNeutralColorTokensAsArray](https://github.com/microsoft/fluentui-react-native/pull/2538/commits/6a9cf328df806e571a3a8d463f25617f43a5fc8d) | failed |

samuelfreiberg commented:
My concern here about the E2E tests is that we're not getting an error screenshot in the dump. A screenshot would tell us a lot here. Going to do a quick investigation

lyzhan7 commented:
> My concern here about the E2E tests is that we're not getting an error screenshot in the dump. A screenshot would tell us a lot here. Going to do a quick investigation

Okay sounds good, thanks Sammy.

lyzhan7 commented:
Have mostly narrowed the iOS e2e test failure issue to this line: https://github.com/microsoft/fluentui-react-native/pull/2538/files/4684a370d0ed9b35b2f7d0f000532bc1cba9e5b5..289f4828829a9cba4d7c532f91de71bc33d8f177

iOS e2e test fails before this line is changed, passes after the change. 
The difference before/after this change is whether we render all 500+ global shared colors vs. just the first one. 
I'll try maybe putting all the global shared color tokens into a collapsible section - if anyone has any other ideas let me know

samuelfreiberg commented:
> Have mostly narrowed the iOS e2e test failure issue to this line: https://github.com/microsoft/fluentui-react-native/pull/2538/files/4684a370d0ed9b35b2f7d0f000532bc1cba9e5b5..289f4828829a9cba4d7c532f91de71bc33d8f177
> 
> iOS e2e test fails before this line is changed, passes after the change. The difference before/after this change is whether we render all 500+ global shared colors vs. just the first one. I'll try maybe putting all the global shared color tokens into a collapsible section - if anyone has any other ideas let me know

Does this change the rendering of the page at all?

lyzhan7 commented:
> > Have mostly narrowed the iOS e2e test failure issue to this line: https://github.com/microsoft/fluentui-react-native/pull/2538/files/4684a370d0ed9b35b2f7d0f000532bc1cba9e5b5..289f4828829a9cba4d7c532f91de71bc33d8f177
> > iOS e2e test fails before this line is changed, passes after the change. The difference before/after this change is whether we render all 500+ global shared colors vs. just the first one. I'll try maybe putting all the global shared color tokens into a collapsible section - if anyone has any other ideas let me know
> 
> Does this change the rendering of the page at all?

Yes there's only one `shared` color instead of 500+, here's a video of what it looks like with that line of change (affects the last test section only)


https://user-images.githubusercontent.com/78454019/214923912-76c028d9-afe4-4985-b0cc-32b752e49e64.mp4

lyzhan7 commented:
Put the 500+ global shared color tokens into a collapsible section in order to pass the iOS e2e tests. Cleaned up the code to remove the old attempts at fixing the e2e tests and updated the description above to include new screen recordings/short explanation of why the collapsible section was required

Saadnajmi commented on apps/fluent-tester/src/TestComponents/ColorTokens/ColorTokenTest.tsx:
```diff
@@ -29,52 +32,72 @@ const styles = StyleSheet.create({
+  return <ColorToken key={colorName} colorValue={colorValue} colorName={colorName} />;
+};
+
+/// Alias color tokens
```
I think the `///` is something used by Xcode but not in JS code much?

lyzhan7 replied:
We use this in a couple other places - ex. Color.types.ts. Should I change it to something else?

Saadnajmi replied:
Oh. If it's an existing pattern then nvm.
